### PR TITLE
dai: rudimentary ALH support

### PR DESCRIPTION
**Pull request by slawblauciak**
This change adds rudimentary support for ALH playback and capture.

It required a refactor in DAIs, namely how DMA handshakes and FIFO addresses are obtained.
From now on, DAI drivers shall implement get_handshake and get_fifo ops, which will be used to retrieve this information.

### CODE REVIEW

**Comment by RanderWang** [comment]
@slawblauciak  How about CML ?

**Comment by slawblauciak** [comment]
@RanderWang we don't have any CML platforms yet

File: src/platform/baytrail/dai.c [hunk 1]
```diff
@@ -12,108 +13,86 @@
 #include <sof/ssp.h>
 #include <ipc/dai.h>
 #include <ipc/stream.h>
+#include <ipc/topology.h>
 #include <config.h>
 
-static struct dai ssp[] = {
```
Comment (tlauda) [inline src/platform/baytrail/dai.c]: Could you elaborate on why dynamic allocation is better than static in this case? You still need to hardcode the number of SSPs and manually assign address and irq.

File: src/include/sof/dai.h [hunk 2]
```diff
@@ -102,7 +104,7 @@ struct dai_plat_data {
 	uint32_t base;
 	uint32_t irq;
 	uint32_t flags;
-	struct dai_plat_fifo_data fifo[2];
+	struct dai_plat_fifo_data *fifo;
```
Comment (tlauda) [inline src/include/sof/dai.h]: What is the benefit?

File: src/drivers/intel/cavs/alh.c [hunk 3]
```diff
@@ -8,6 +8,8 @@
 #include <sof/dma.h>
 #include <sof/trace.h>
 #include <ipc/dai.h>
+#include <platform/alh.h>
```
Comment (tlauda) [inline src/drivers/intel/cavs/alh.c]: Include it in sof/alh.h instead and include sof/alh.h here.

File: src/drivers/intel/cavs/alh.c [hunk 4]
```diff
@@ -8,6 +8,8 @@
...
+#include <platform/memory.h>
```
Comment (tlauda) [inline src/drivers/intel/cavs/alh.c]: sof/memory.h

File: src/drivers/intel/cavs/alh.c [hunk 5]
```diff
@@ -59,14 +61,15 @@ static int alh_remove(struct dai *dai)
...
-	/* TODO */
-	return 0;
+	return alh_handshake_map[stream_id];
 }
 
 static int alh_get_fifo(struct dai *dai, int direction, int stream_id)
 {
-	/* TODO */
-	return 0;
+	uint32_t offset = direction == SOF_IPC_STREAM_PLAYBACK ?
```
Comment (tlauda) [inline src/drivers/intel/cavs/alh.c]: #include <stdint.h>

File: src/include/sof/dai.h (hunk 2 above)
Comment (plbossart) [inline src/include/sof/dai.h]: I am not sure this is a required change. this leads to a lot of changes for SSP/DMIC/HDA, will increase code size for no good reason, and you aren't using this field at all for ALH. So why not keep the fifo[2] and just implement get_fifo/handshake in a ALH-specific way when you need them? You would waste a bit of memory that's not used for ALH but it'd be more efficient in the end, no?

File: src/platform/icelake/include/platform/alh.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+
+#ifndef __PLATFORM_ALH_H__
```
Comment (tlauda) [inline src/platform/icelake/include/platform/alh.h]: Add #ifdef __SOF_ALH_H__ to force inclusion only from sof/alh.h. See other platform headers.

File: src/platform/icelake/include/platform/alh.h [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ */
+const uint8_t alh_handshake_map[] = {
+	-1,	/* 0  */
+	-1,	/* 1  */
+	-1,	/* 2  */
+	-1,	/* 3  */
+	-1,	/* 4  */
+	-1,	/* 5  */
+	-1,	/* 6  */
+	22,	/* 7  */
```
Comment (tlauda) [inline src/platform/icelake/include/platform/alh.h]: Could you include descriptions of used handshakes?
Or maybe even make them defines, so they would describe themselves.

File: src/include/sof/dai.h (hunk 2 above)
Comment (q009) [reply src/include/sof/dai.h]: This was my thought, initially I thought about allocating a larger array here for ALH, but I wasn't the fan having such large structures on stack, hence this. So yeah, I will revert to using 2 default fifo structures.

File: src/platform/icelake/include/platform/alh.h (hunk 7 above)
Comment (q009) [reply src/platform/icelake/include/platform/alh.h]: The firmware would never use those handshakes in the code directly, so the defines would remain unused. We just retrieve the right handshake by using the stream_id we get from the driver.

Comment (tlauda) [reply src/platform/icelake/include/platform/alh.h]: OK, so maybe at least description in the comment?

File: src/platform/icelake/include/platform/memory.h [hunk 8]
```diff
@@ -101,6 +101,12 @@
 #define SSP_BASE(x)		(0x00077000 + x * SSP_SIZE)
 #define SSP_SIZE		0x0000200
 
+/* ALH */
+#define ALH_BASE		0x000071000
+#define ALH_TXDA_OFFSET		0x000000400
+#define ALH_RXDA_OFFSET		0x000000500
+#define ALH_STREAM_OFFSET	0x000000004
+
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/platform/icelake/include/platform/memory.h]: I ported this to CML and made soundwire work on CML.

File: src/platform/icelake/include/platform/alh.h (hunk 6 above)
Comment (slawblauciak) [reply src/platform/icelake/include/platform/alh.h]: I decided to throw an error if included in a wrong place instead, that way the users will not get the silent treatment.

**Comment by slawblauciak** [comment]
Changes applied.

**Comment by slawblauciak** [comment]
SOFCI TEST

File: src/drivers/intel/cavs/alh.c [hunk 9]
```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+
+#include <stdint.h>
```
Comment (tlauda) [inline src/drivers/intel/cavs/alh.c]: The rule is the most global headers at the end and also in alphabetical order across their own group:
#include <sof/alh.h>
#include <sof/dai.h>
#include <sof/dma.h>
#include <sof/memory.h>
#include <sof/trace.h>
#include <ipc/dai.h>
#include <stdint.h>

File: src/drivers/intel/cavs/alh.c [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+static int alh_get_handshake(struct dai *dai, int direction, int stream_id)
+{
+	return alh_handshake_map[stream_id];
+}
+
+static int alh_get_fifo(struct dai *dai, int direction, int stream_id)
+{
+	uint32_t offset = direction == SOF_IPC_STREAM_PLAYBACK ?
```
Comment (tlauda) [inline src/drivers/intel/cavs/alh.c]: <ipc/stream.h>

File: src/include/sof/alh.h [hunk 11]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+
+#ifndef __INCLUDE_ALH__
```
Comment (tlauda) [inline src/include/sof/alh.h]: Should go to sof/drivers/alh.h.
And also look at all other guards: `__SOF_DRIVERS_ALH_H__`

File: src/include/sof/alh.h [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+#define __INCLUDE_ALH__
+
+#define SOF_ALH_H
+
+#include <sof/dai.h>
+#include <platform/alh.h>
```
Comment (tlauda) [inline src/include/sof/alh.h]: #include <platform/alh.h>
#include <sof/dai.h>

File: src/platform/icelake/include/platform/alh.h [hunk 13]
```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Comment (tlauda) [inline src/platform/icelake/include/platform/alh.h]: include/platform/drivers/alh.h

File: src/platform/icelake/include/platform/memory.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply src/platform/icelake/include/platform/memory.h]: @RanderWang can you share your changes for CML?

File: src/include/sof/alh.h [hunk 14]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+
+#ifndef __INCLUDE_ALH__
+#define __INCLUDE_ALH__
+
+#define SOF_ALH_H
```
Comment (tlauda) [inline src/include/sof/alh.h]: Could we start a naming pattern here and maybe name it exactly like guard, but without underscores?
`#define SOF_DRIVERS_ALH_H`

File: src/platform/icelake/include/platform/alh.h [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+
+#ifndef SOF_ALH_H
```
Comment (tlauda) [inline src/platform/icelake/include/platform/alh.h]: Your solution with undef is even better than my initial idea, but could we follow already implemented pattern? See e.g. include/platform/lib/clk.h. I would like to be consistent across the whole source code, so:
```
#if SOF_DRIVERS_ALH_H

#ifndef __PLATFORM_DRIVERS_ALH_H__
#define __PLATFORM_DRIVERS_ALH_H__

<code here>

#endif /* __PLATFORM_DRIVERS_ALH_H__ */

#else 

#error "This file shouldn't be included from outside of sof/drivers/alh.h"

#endif /* SOF_DRIVERS_ALH_H */
```

File: src/platform/icelake/include/platform/alh.h (hunk 13 above)
Comment (slawblauciak) [reply src/platform/icelake/include/platform/alh.h]: This is an icelake specific header

Comment (tlauda) [reply src/platform/icelake/include/platform/alh.h]: Yes, exactly what I meant: src/platform/icelake/include/platform/drivers/alh.h (see latest dir structure for reference).

**Comment by slawblauciak** [comment]
Changes applied

File: src/platform/icelake/include/platform/memory.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/platform/icelake/include/platform/memory.h]: @slawblauciak  I added these code to cannonlake to support soundwire, could you please also apply this change ? And also please add alh.h to cannonlake platform. Thanks!

diff --git a/src/platform/cannonlake/include/platform/dai.h b/src/platform/cannonlake/include/platform/dai.h

@@ -32,6 +32,9 @@
 /** \brief Number of HD/A Link Inputs */
 #define DAI_NUM_HDA_IN		7
 
+/** \brief Number of ALH bi-directional links */
+#define DAI_NUM_ALH_BI_DIR_LINKS	16
+
 int dai_init(void);
 
 #endif
diff --git a/src/platform/cannonlake/include/platform/memory.h b/src/platform/cannonlake/include/platform/memory.h

@@ -57,6 +57,12 @@
 #define MN_BASE			0x00078C00
 #define MN_SIZE			0x00000200
 
+/* ALH */
+#define ALH_BASE		0x000071000
+#define ALH_TXDA_OFFSET		0x000000400
+#define ALH_RXDA_OFFSET		0x000000500
+#define ALH_STREAM_OFFSET	0x000000004
+
 /* low power DMA position */
 #define LP_GP_DMA_LINK_SIZE	0x00000010
 #define LP_GP_DMA_LINK_BASE(x) (0x00001C00 + x * LP_GP_DMA_LINK_SIZE)

[the remark this answers is not in the document]
Comment (slawblauciak) [reply src/platform/icelake/include/platform/memory.h]: Very well, thanks

**Comment by tlauda** [comment]
@slawblauciak Could you link the kernel patch?

**Comment by slawblauciak** [comment]
@tlauda I don't think the kernel patch is ready. I might be wrong though...
@plbossart @RanderWang

**Comment by RanderWang** [comment]
> @tlauda I don't think the kernel patch is ready. I might be wrong though...
> @plbossart @RanderWang

As I know, kernel patch is not ready, we are still refining it.

**Comment by plbossart** [comment]
> > @tlauda I don't think the kernel patch is ready. I might be wrong though...
> > @plbossart @RanderWang
> 
> As I know, kernel patch is not ready, we are refining it.



> > @tlauda I don't think the kernel patch is ready. I might be wrong though...
> > @plbossart @RanderWang
> 
> As I know, kernel patch is not ready, we are refining it.

yep, still WIP and we have dependencies on pushing code through the SoundWire tree first, ETA ~2 weeks. Please go ahead on the firmware side.

File: src/audio/dai.c [hunk 16]
```diff
@@ -728,10 +757,30 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
...
+		 */
+		channel = config->alh.stream_id;
+		dd->stream_id = config->alh.stream_id;
+		trace_dai_with_ids(dev, "dai_config(), channel = %d",
+				   channel);
+
+		if (dd->chan != DMA_CHAN_INVALID) {
+			dma_channel_put(dd->dma, dd->chan);
+			dd->chan = DMA_CHAN_INVALID;
+		}
```
Comment (RanderWang) [inline src/audio/dai.c]: hi @slawblauciak  please delete the following line
````
		if (dd->chan != DMA_CHAN_INVALID) {
			dma_channel_put(dd->dma, dd->chan);
			dd->chan = DMA_CHAN_INVALID;
		}
````
Now sdw doesn't work with pulse audio and the reason is that: dai_config will be triggered for both playback and
capture to free dma channel for they use the same dai_index(please check https://github.com/thesofproject/sof/blob/master/src/ipc/ipc.c#L406). For HDA, playback and capture use different dai_index. So if both playback & capture are working, and playback is stopped then capture failed.

Alh uses dw-dma, not link-dma. For link-dma, the dma channel is specialized by kernel driver,
but there is no such requirement for dw-dma. Actually ahl is closer to I2S than HDA. please check there is no
such operation for I2S(SSP).

**Comment by slawblauciak** [comment]
Changes applied

**Comment by slawblauciak** [comment]
Made a PR on the driver side to align the ABI with this change.

**Comment by lgirdwood** [comment]
@slawblauciak looks like the CI needs rerun here and can you reply with kernel PR number so it's easy to track.

**Comment by slawblauciak** [comment]
Kernel change merged: https://github.com/thesofproject/linux/pull/1129

**Comment by ranj063** [comment]
@slawblauciak could you please rebase this to master so we can test this on TGL?

**Comment by plbossart** [comment]
> @slawblauciak could you please rebase this to master so we can test this on TGL?

@ranj063 use the sdw-001-drop-stable branch for now.

File: src/drivers/intel/cavs/alh.c [hunk 17]
```diff
@@ -63,14 +64,15 @@ static int alh_remove(struct dai *dai)
 
 static int alh_get_handshake(struct dai *dai, int direction, int stream_id)
 {
-	/* TODO */
-	return 0;
+	return alh_handshake_map[stream_id];
```
Comment (lgirdwood) [inline src/drivers/intel/cavs/alh.c]: Is stream_id passed in from host ? Should we validate it's value before any array lookup ?

File: src/drivers/intel/cavs/Kconfig [hunk 18]
```diff
@@ -103,9 +103,9 @@ config CAVS_SSP
...
 
-config CAVS_SOUNDWIRE
-	bool "Intel cAVS SoundWire driver"
+config CAVS_ALH
+	bool "Intel cAVS ALH driver"
 	depends on CAVS
 	default n
 	help
-	  Select this to enable Intel cAVS SoundWire driver.
+	  Select this to enable Intel cAVS ALH driver.
```
[the remark this answers is not in the document]
Comment (slawblauciak) [reply src/drivers/intel/cavs/Kconfig]: There's no explicit SoundWire support in the firmware, only ALH. It is not covered in this PR.

**Comment by slawblauciak** [comment]
Sync'd with sdw-001-drop-stable branch

File: src/arch/xtensa/configs/icelake_defconfig [hunk 19]
```diff
@@ -1,6 +1,7 @@
 CONFIG_ICELAKE=y
 CONFIG_CAVS_DMIC=y
 CONFIG_CAVS_SSP=y
+CONFIG_CAVS_ALH=y
```
Comment (jajanusz) [inline src/arch/xtensa/configs/icelake_defconfig]: Please add it to tigerlake

File: src/platform/icelake/include/platform/drivers/alh.h [hunk 20]
```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Comment (jajanusz) [inline src/platform/icelake/include/platform/drivers/alh.h]: I guess we need it also for tgl

File: src/arch/xtensa/configs/icelake_defconfig (hunk 19 above)
Comment (slawblauciak) [reply src/arch/xtensa/configs/icelake_defconfig]: Will come in a separate PR

File: src/platform/icelake/include/platform/drivers/alh.h (hunk 20 above)
Comment (slawblauciak) [reply src/platform/icelake/include/platform/drivers/alh.h]: In a separate PR, likely requires different handshakes.

**Comment by plbossart** [comment]
@slawblauciak I was expecting some WIP patches to be squashed and the hacks removed?

**Comment by slawblauciak** [comment]
@plbossart yeah, I suppose it is a good idea, seeing the amount of commits

**Comment by slawblauciak** [comment]
@plbossart I'm not sure what to do about the topology hacks

**Comment by plbossart** [comment]
> @plbossart I'm not sure what to do about the topology hacks
> Should the HDMI remain left out?

I suspect we'd want to keep two versions, one with and one without HDMI since it's still quite flaky on ICL

File: src/audio/dai.c [hunk 21]
```diff
@@ -769,10 +769,31 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
...
 		break;
+	case SOF_DAI_INTEL_ALH:
+		/* set to some non-zero value to satisfy the condition below,
+		 * it is recalculated in dai_params() later
+		 */
+		dd->frame_bytes = 4;
+
+		/* SDW HW FIFO requires 32bit MSB aligned sample data for
+		 * all formats, such as 8/16/24/32 bits. Now set frame_fmt
+		 * to S32_LE to let volume component do the conversion.
```
Comment (plbossart) [inline src/audio/dai.c]: the comment is a bit misleading, the dai should not assume that there's a volume component but only that it deals with data that has been left-aligned, and throw an error if this is not the case,

**Comment by slawblauciak** [comment]
Changes applied. Also merged changes from #1795

**Comment by slawblauciak** [comment]
@plbossart already on it :)

**Pull request merged**